pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -269,14 +269,21 @@ func (rc *reconciler) reconcile(ctx context.Context) {
 			verifySafeToDetach := !(timeout || hasOutOfServiceTaint)
 			err = rc.attacherDetacher.DetachVolume(logger, attachedVolume.AttachedVolume, verifySafeToDetach, rc.actualStateOfWorld)
 			if err == nil {
-				if !timeout {
+				if verifySafeToDetach { // normal detach
 					logger.Info("attacherDetacher.DetachVolume started", "node", klog.KRef("", string(attachedVolume.NodeName)), "volumeName", attachedVolume.VolumeName)
-				} else {
-					metrics.RecordForcedDetachMetric()
-					logger.Info("attacherDetacher.DetachVolume started: this volume is not safe to detach, but maxWaitForUnmountDuration expired, force detaching",
-						"duration", rc.maxWaitForUnmountDuration,
-						"node", klog.KRef("", string(attachedVolume.NodeName)),
-						"volumeName", attachedVolume.VolumeName)
+				} else { // force detach
+					if timeout {
+						metrics.RecordForcedDetachMetric(metrics.ForceDetachReasonTimeout)
+						logger.Info("attacherDetacher.DetachVolume started: this volume is not safe to detach, but maxWaitForUnmountDuration expired, force detaching",
+							"duration", rc.maxWaitForUnmountDuration,
+							"node", klog.KRef("", string(attachedVolume.NodeName)),
+							"volumeName", attachedVolume.VolumeName)
+					} else {
+						metrics.RecordForcedDetachMetric(metrics.ForceDetachReasonOutOfService)
```

Comment:
This is slightly different behavior---we didn't log on the out of service taint before?

Regardless, seems like a good idea to do this.

Reply:
Yes. When trying to add reason for the forced detach, I noticed that this logic should change a little bit so that forced detach due to out-of-service taint should get a different logging message.